### PR TITLE
avoid throwing for flow control

### PR DESCRIPTION
**Pull request by charlierudolph**
Instead of having a try/catch around this function just have it return an optional error message.

### CODE REVIEW

File: src/cli/assert_node_engine_version_spec.ts [hunk 1]
```diff
@@ -1,18 +1,33 @@
...
-    assertNodeEngineVersion('v17.0.0')
+  it('returns null when the version is greater than specified in package.json', () => {
+    // Arrange
+
+    // Act
+    const result = validateNodeEngineVersion('v17.0.0', () => ({
+      engines: {
+        node: '>=12',
+      },
+    }))
```
Comment (charlierudolph) [inline src/cli/assert_node_engine_version_spec.ts:28]: Think its better to hardcode the engines here as well so this doesn't suddenly fail when we no longer support version 17 (probably a little ways off)

**Review by aurelien-reeves — CHANGES_REQUESTED** [review]
Avoiding throwing an error can be a good idea 👍

But rather than returning `string | null` - which actually I would have expect to be the return type of the validateNodeEngineVersion  function- with the string being kinda error message, would it be possible to return something that would look like a result struct which could looks like the following?

```typescript
{
  status: bool,
  message: string?
}
```

Or, just, a boolean, and the caller could be responsible for the error message?

File: src/cli/assert_node_engine_version_spec.ts (hunk 1 above)
Comment (mattwynne) [reply src/cli/assert_node_engine_version_spec.ts:30]: 💯, I missed that.

**Comment by mattwynne** [comment]
I did consider using a promise as a kind of Maybe object, so the calling code could do something like

```
validateNodeEngineVersion(process.version).catch(errorMessage => {
  console.error(error) // eslint-disable-line no-console
  process.exit(1)
})
```

Still using exceptions for flow control really, but perhaps it reads better?

Another alternative would be to pass the assertion function a callback to run if the engine is invalid, like:

```
validateNodeEngineVersion(process.version, errorMessage => {
  console.error(error) // eslint-disable-line no-console
  process.exit(1)
})
```

That way the conditional stays inside the function. That's probably my favourite of them all.

File: src/cli/assert_node_engine_version_spec.ts (hunk 1 above)
Comment (davidjgoss) [reply src/cli/assert_node_engine_version_spec.ts:30]: Oh true, the list right now is 12, 14, 16, 17

**Comment by davidjgoss** [comment]
@mattwynne I quite like the callback idea there.

**Review by davidjgoss — APPROVED** [review]

**Review by mattwynne — APPROVED** [review]
lovely!

File: src/cli/assert_node_engine_version_spec.ts [hunk 2]
```diff
@@ -1,18 +1,37 @@
...
+
+    // Assert
+    expect(errorSpy).to.have.been.calledOnceWith(
+      'Cucumber can only run on Node.js versions >=12. This Node.js version is v11.0.0'
     )
   })
 
-  it('passes when the version is greater than specified in package.json', () => {
-    assertNodeEngineVersion('v17.0.0')
+  it('returns null when the version is greater than specified in package.json', () => {
```
Comment (mattwynne) [inline src/cli/assert_node_engine_version_spec.ts:23]: I just realised we should have changed these example descriptions too: we're not returning anything anymore.

**Pull request merged**